### ISSUE TITLE
Randomly failing unittest: test_default_configuration_classify

### ISSUE THREAD

**Issue by Ashafix** [automl/auto-sklearn#734]
The test method `test_default_configuration_classify` in `test_kernel_pca.py` randomly fails.

**test.py**
```
from test.test_pipeline.components.feature_preprocessing.test_kernel_pca import KernelPCAComponentTest
KernelPCAComponentTest().test_default_configuration_classify()
```

**test.sh**
```
#!/bin/bash

status=0
while [ "$status" -eq 0 ]; do
  python3 test.py
  status=$?
  echo "$status"
done
```

Most times no error is thrown but sometimes it fails with `RuntimeWarning: invalid value encountered in multiply`, e.g. here https://travis-ci.org/automl/auto-sklearn/jobs/600007367 or https://travis-ci.org/automl/auto-sklearn/jobs/599956798

From the code I cannot tell where the random magic happens, `random_state` is set and even using the same `X_train` and `Y_train` cannot reproduce the error.

**Comment by mfeurer** [comment]
Thanks for pointing this out, this is indeed weird. Would you be able to extract this error from Auto-sklearn to produce a minimal example which could be sent to scikit-learn?

**Comment by Ashafix** [comment]
It's a truly weird bug, totally sensitive to minor changes in the code, execution order, meaningless `print` statments, how it's called in shell scripts, pyc-cache files, etc.

* moving the `import` statements in `KernelPCA.fit` to the top of the module, error is gone
* changing `copy` `check_array` to `False`, error is gone (but opposite to the stand-alone snippet `test.py`)

For a demonstration and a minimal sklearn example see the following snippets.

```
find . -name *.pyc -delete
python3 test_minimal.py
python3 test_minimal.py
```
Crashes the first time, but not the second time it is called.

```
find . -name *.pyc -delete
python3 test.py
```
gives a warning but no error. Now move the `fit` statement inside the `warning` block (as in the auto-learn code) and I gives the same error (as test_minimal.py) but the error stays after calling the program more than once.

Now for some fun add the following import statements to `test_minimal.py`
```
import scipy.sparse
import sklearn.decomposition
```
delete the pyc files (as before) and run the program twice, it crashes twice.

Now for some extra fun, remove the above import statements from the `fit` function in `KernelPCA` and add it to the beginning of the module. After removing the pyc files we can call the script `test_minimal.py` without errors.

I don't know if my system is screwed up (tried a clean virtualenv with similar behavior) or what is going on here.

 


OS: Ubuntu 18.04 LTS
Python: 3.6.5







**test.sh**
```

#!/bin/bash

find . -name *.pyc -delete

status=0

while [ "$status" -le 1 ]; do

  python3 test.py
  status=$((status + $?))
  echo "exit code: $status"

  find . -name *.pyc -delete
  python3 test_minimal.py
  status=$((status + $?))
  echo "exit code: $status"
done
```

**test_minimal.py**
```
from test.test_pipeline.components.feature_preprocessing.test_kernel_pca import KernelPCAComponentTest

KernelPCAComponentTest().test_default_configuration_classify()
```

**test.py**
```
import numpy as np
import sklearn.metrics
import sklearn.decomposition
from sklearn.utils.validation import check_array
from sklearn.preprocessing import KernelCenterer
import sklearn.datasets
import warnings


def test_default_configuration_classify():
    dataset = sklearn.datasets.load_digits()
    X = dataset.data.astype(np.float32)
    rs = np.random.RandomState(42)
    indices = np.arange(X.shape[0])
    train_size = min(int(len(indices) / 3. * 2.), 1000)
    rs.shuffle(indices)
    X_train = X[indices][:train_size]

    for i in range(5):
        print('i: {}'.format(i))

        X = check_array(X_train, accept_sparse='csr', copy=False)
        preprocessor = sklearn.decomposition.KernelPCA(n_components=100, kernel='rbf',
                                                       degree=3, gamma=1.0, coef0=0.0,
                                                       remove_zero_eig=True, random_state=1)

        preprocessor._centerer = KernelCenterer()
        K = preprocessor._get_kernel(X)
        with warnings.catch_warnings():
            warnings.filterwarnings("error")
        preprocessor._fit_transform(K)

resp = test_default_configuration_classify()
```

**Comment by gui-miotto** [comment]
Hi @Ashafix , thanks for investigating this issue.
By the way, I think this problem is an old friend auto-sklearn. I call it the "**np.newaxis thing**". But I never bother investigating it because, as you pointed out, the error occurs stochastically. Usually we "get around it" just by re-running the test.
But today I looked at it more carefully and I think I understand now what is causing the problem.

I suspect the problem is on scikit-learn, more preciselly on these lines of ```_kernel_pca.py```:
```python
        # flip eigenvectors' sign to enforce deterministic output
        self.alphas_, _ = svd_flip(self.alphas_,
                                   np.empty_like(self.alphas_).T)
```
https://github.com/scikit-learn/scikit-learn/blob/b853653234b6b0c15b2d2501db2e72041005f813/sklearn/decomposition/_kernel_pca.py#L217-L219

**TLDR**: Replace ```np.empty_like()``` by  ```np.ones_like()``` and the problem is fixed.

Why?

The function  ```svd_flip``` is meant to transform two matrices and return the result of these two transformations. But here, as you can see in the code snippet, we just care about one matrix:  ```self.alphas_ ```. So they pass as a 2nd parameter a junk matrix built with ```np.empty_like()```. This matrix is transformed, but its result is ignored (attributed to  ```_ ```).

(Here is the code for ```svd_flip```:)
https://github.com/scikit-learn/scikit-learn/blob/b853653234b6b0c15b2d2501db2e72041005f813/sklearn/utils/extmath.py#L494-L535

Ok, but the problem is that ```np.empty()``` can build a matrix with numbers that are not allowed according to the **IEEE754**. So one should not use a matrix build with ```np.empty()``` straightaway on arithmetic. However that is exactly what is happening here. The function  ```svd_flip``` gets this junk matrix and multiplies it by another one.  This is the line where this happens:

```python
v *= signs[:, np.newaxis]
```
https://github.com/scikit-learn/scikit-learn/blob/b853653234b6b0c15b2d2501db2e72041005f813/sklearn/utils/extmath.py#L528

So, to fix this I substituted ```np.empty_like()``` by ```np.zeros_like()``` or ```np.ones_like()``` and got no more errors.

@Ashafix , It would be great if you could try this fix as well and give us a feedback.

Cheers

**Comment by gui-miotto** [comment]
A possible fix has just been merged into the master of scikit-learn. Therefore, I suggest closing this issue for now.
https://github.com/scikit-learn/scikit-learn/pull/16718
